#!/usr/bin/env cwl-runner

# (Re)generated by BlueBee Platform

$namespaces:
  bb: http://bluebee.com/cwl/
  ilmn-tes: http://platform.illumina.com/rdf/iap/
cwlVersion: cwl:v1.0
class: CommandLineTool
bb:toolVersion: ''
id: filter_blastn
requirements:
- class: InitialWorkDirRequirement
  listing:
  - entry: |-
      #!/bin/bash
      set -euo pipefail
      awk  -v x="/"  'BEGIN {FS=OFS="\t"} {split($1, a, x);  print a[1], $1, $3}' $1 | uniq > tmp1
      awk  -v x="/"  'BEGIN {FS=OFS="\t"} {split($1, a, x);  print a[1], $1, $2}' $2 | uniq > tmp2
      awk  'BEGIN {FS=OFS="\t"} NR==FNR{a[$1]=$1;next} a[$1]!=$1{print $2,$3}' tmp1 tmp2 | tee filtered | awk 'BEGIN {FS=OFS="\t"} {print $1}' > ids.txt
      wc -l filtered > NM_paired.number
      awk -F"\t" '{print ">"$1"\n"$2}' filtered > NM_paired_filtered.fasta
    entryname: filter_blastn.sh
    writable: false
hints:
- class: DockerRequirement
  dockerPull: clinical_metagenomics_preprocessing:latest
- class: ResourceRequirement
  ilmn-tes:resources:
    tier: standard
    type: standardHiMem
    size: large
label: filter_blastn
stderr: stderr.txt
inputs:
  blastn_result:
    type: File
    inputBinding:
      position: 1
  blast_query:
    type: File
    inputBinding:
      position: 2
outputs:
  filtered_blastn_hits_fasta:
    type: File
    outputBinding:
      glob:
      - '*_filtered.fasta'
  count:
    type: File
    outputBinding:
      glob:
      - '*.number'
  ids:
    type: File
    outputBinding:
      glob:
      - ids.txt
  error_log:
    type: stderr
  filter_blastn_script:
    type: File
    outputBinding:
      glob:
      - filter_blastn.sh
  paired_filtered:
    type: File
    outputBinding:
      glob:
      - filtered
baseCommand:
- bash
- filter_blastn.sh
