#!/usr/bin/env cwl-runner

# (Re)generated by BlueBee Platform

$namespaces:
  bb: http://bluebee.com/cwl/
  ilmn-tes: http://platform.illumina.com/rdf/iap/
cwlVersion: cwl:v1.0
class: CommandLineTool
bb:toolVersion: '1'
requirements:
- class: InitialWorkDirRequirement
  listing:
  - entry: $(inputs.blast_output)
    entryname: blastoutput.tab
    writable: false
  - entry: $(inputs.read_lengths)
    entryname: read_lenghts.tsv
    writable: false
  - entry: $(inputs.accession2taxid)
    entryname: accession2taxid
    writable: false
  - entry: $(inputs.nuclLength)
    entryname: nuclLength.tab
    writable: false
  - entry: $(inputs.taxonomy_names)
    entryname: names.dmp
    writable: false
  - entry: |-
      library(metaMix)
      step1<-generative.prob.nucl(blast.output.file="blastoutput.tab",  outDir="./", blast.default=TRUE, read.length.file="read_lenghts.tsv", contig.weight.file=1,  accession.taxon.file="accession2taxid", genomeLength="nuclLength.tab")
      step2<-reduce.space(step1="step1.RData", taxon.name.map="names.dmp")
      step3<-parallel.temper.nucl(step2="step2.RData", readSupport=10)
      step4<-bayes.model.aver(step2="step2.RData", step3="step3.RData", taxon.name.map="names.dmp")
    entryname: run_metamix.R
    writable: false
  - entry: |-
      #!/bin/bash
      set -euo pipefail
      echo -n "localhost slots=" > ./hostfile
      nproc --all >> ./hostfile
      mpirun $@
    entryname: run.sh
    writable: false
hints:
- class: DockerRequirement
  dockerPull: metamix:latest
- class: ResourceRequirement
  ilmn-tes:resources:
    tier: standard
    type: standardHiMem
    size: large
label: metamix
inputs:
  blast_output:
    type: File
  read_lengths:
    type: File
  accession2taxid:
    type: File
  nuclLength:
    type: File
  taxonomy_names:
    type: File
outputs:
  allreads_classified:
    type: File
    outputBinding:
      glob:
      - allreads_classified.tab
  histograms_cdf:
    type: File
    outputBinding:
      glob:
      - histograms_cdf.pdf
  logLikelihood_traceplot_40:
    type: File
    outputBinding:
      glob:
      - logLikelihood_traceplot_40.pdf
  logLikelihood_traceplot_all:
    type: File
    outputBinding:
      glob:
      - logLikelihood_traceplot_all.pdf
  presentSpecies_assignedReads:
    type: File
    outputBinding:
      glob:
      - presentSpecies_assignedReads.tsv
  step2_species_annotated:
    type: File
    outputBinding:
      glob:
      - step2_species_annotated.tsv
arguments:
- position: 1
  prefix: --hostfile
  valueFrom: hostfile
- position: 2
  prefix: -np
  valueFrom: '1'
- position: 3
  valueFrom: Rscript
- position: 4
  valueFrom: ./run_metamix.R
- position: 5
  valueFrom: step.out
baseCommand:
- bash
- run.sh
